core/src/main/java/com/datastax/oss/driver/internal/core/CqlIdentifiers.java

```diff
@@ -16,21 +16,36 @@
 package com.datastax.oss.driver.internal.core;
 
 import com.datastax.oss.driver.api.core.CqlIdentifier;
+import com.datastax.oss.driver.shaded.guava.common.base.Preconditions;
 import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableList;
 import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableMap;
+import edu.umd.cs.findbugs.annotations.NonNull;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Function;
 
 public class CqlIdentifiers {
 
-  public static List<CqlIdentifier> wrap(Iterable<String> in) {
+  public static List<CqlIdentifier> wrap(
+      @NonNull Iterable<String> in, @NonNull Function<String, CqlIdentifier> fn) {
+
+    Preconditions.checkNotNull(in, "Input Iterable must not be null");
+    Preconditions.checkNotNull(fn, "CqlIdentifier conversion function must not be null");
     ImmutableList.Builder<CqlIdentifier> out = ImmutableList.builder();
```

Comment:
Our contribution guidelines forbid the usage of Streams API but I often think that this is too excessive. Here, for instance, one line of code could do the job, and it's very readable (and technically speaking, this _is_ a stream transformation):

```
Streams.stream(in).map(fn).collect(ImmutableList.toImmutableList());
```

If I'm not the only one thinking so, we can relax the guidelines a bit for cases like this.

Reply:
I disagree. Examples like this are what I had in mind when I wrote the guidelines.

The goal of this method is to transform a handful of identifiers. Looking at the call sites, it will probably be less than 5 most of the time. It doesn't need "a pipeline of computational operations, functional in nature, laziness-seeking and possibly unbounded".

Readability: not much of a factor because this is a trivial example, any version would be pretty easy to figure out. Yet you can't really beat "for each element, apply the function".

Performance: again it's probably very close, and it's not on the hot path so a few nanoseconds won't make a difference. But conceptually there's something about "convert to a stream, do operations and convert back to a list" that doesn't feel right for 5 elements.

One-liner: yeah it crams everything on one line, is that a good thing? The imperative version is 5 lines, probably doesn't take very long for your brain to parse either. And the typing time is probably very close too, if that even matters.


I'm not against the stream API if we find the need to apply computational steps to a large stream of elements. But I don't get what it buys us over a plain old for loop for trivial operations on small collections.

Reply:
There's clearly not a consensus on any change to the coding guidelines so for now I'm leaving this as-is.  I do very much agree that this is a debate worth having, although this PR is not the place for it... especially given the time-sensitive nature of this ticket. :)

Reply:
And I disagree with the disagreement.